### PR TITLE
Support distributed training

### PR DESCRIPTION
**Pull request by tomaszpietruszka-globality**
This PR is meant to enable training models in a "distributed" manner (even if within the same instance) - with multiple processes, where only the "master" process should communicate with the outside world (filesystem, wandb, ...)

If we detect that the current process is a member of a distributed cluster - and it is not the master process - it will refrain from :
- saving artifacts
- interacting with "Metrics observers" (such as wandb)

In all other cases, nothing changes.

Additionally, added a CI stage to publish Release Candidates to JFrog (based on https://github.com/globality-corp/microcosm-flask/pull/319)

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -211,9 +211,25 @@ jobs:
       - attach_workspace:
           at: ~/repo
       - run:
-          name: Deploy
+          name: Override version
```
Comment (uri-miron-globality) [inline .circleci/config.yml:214]: Is there a change needed in build.json to prevent this from being overwritten the next time globality-build is run?

Comment (marcglobality) [reply .circleci/config.yml:214]: @tomaszpietruszka-globality ?

File: microcosm_sagemaker/bundle_traversal.py [hunk 2]
```diff
@@ -53,18 +53,17 @@ def fit_and_save_bundle(
...
     bundle_orchestrator="single_threaded_bundle_orchestrator",
 )
 class BundleAndDependenciesTrainer:
     def __init__(self, graph: ObjectGraph):
-        self.bundle_orchestrator = getattr(
-            graph,
-            graph.config.bundle_and_dependencies_trainer.bundle_orchestrator,
-        )
+        config = graph.config.bundle_and_dependencies_trainer
+        self.bundle_orchestrator = getattr(graph, config.bundle_orchestrator)
```
Comment (marcglobality) [inline microcosm_sagemaker/bundle_traversal.py:66]: I know for you it's obvious, but can you help the noob reader a bit here? what values can `bundle_orchestrator` take? `single_threaded_bundle_orchestrator` and `distributed_cluster` ? could you write a comment to explain?

File: .circleci/config.yml (hunk 1 above)
Comment (tomaszpietruszka-globality) [reply .circleci/config.yml]: There was no build.json change necessary - but an update to `globality-build` was -> https://github.com/globality-corp/globality-build/pull/599

File: microcosm_sagemaker/bundle_traversal.py [hunk 3]
```diff
@@ -14,10 +16,8 @@
 )
 class BundleAndDependenciesLoader:
     def __init__(self, graph: ObjectGraph):
-        self.bundle_orchestrator = getattr(
-            graph,
-            graph.config.bundle_and_dependencies_loader.bundle_orchestrator,
-        )
+        config = graph.config.bundle_and_dependencies_loader
+        self.bundle_orchestrator: BundleOrchestrator = getattr(graph, config.bundle_orchestrator)
```
Comment (tomaszpietruszka-globality) [inline microcosm_sagemaker/bundle_traversal.py:20]: unrelated readability improvement

File: microcosm_sagemaker/bundle_traversal.py [hunk 4]
```diff
@@ -53,18 +53,17 @@ def fit_and_save_bundle(
...
     bundle_orchestrator="single_threaded_bundle_orchestrator",
 )
 class BundleAndDependenciesTrainer:
     def __init__(self, graph: ObjectGraph):
-        self.bundle_orchestrator = getattr(
-            graph,
-            graph.config.bundle_and_dependencies_trainer.bundle_orchestrator,
-        )
+        config = graph.config.bundle_and_dependencies_trainer
+        self.bundle_orchestrator: BundleOrchestrator = getattr(graph, config.bundle_orchestrator)
```
Comment (tomaszpietruszka-globality) [inline microcosm_sagemaker/bundle_traversal.py:66]: unrelated readability improvement

**Review by marcglobality — APPROVED** [review]
LGTM, thanks for the initiative!

**Comment by aimore-globality** [comment]
That seems ok to me. I would love to see it practice!

**Review by aimore-globality — APPROVED** [review]

**Pull request merged**